tenable/io/v3/schema.py

```diff
@@ -1,13 +1,82 @@
 '''
-Users API Endpoints Schemas
+V3 API Endpoints Schemas
 '''
-from marshmallow import Schema, fields
+from marshmallow import Schema, ValidationError, fields
+from marshmallow import validate as v
+from marshmallow.decorators import post_dump, pre_load
+
+
+class AssignTagsAssetSchema(Schema):
+    '''
+    Assign Tags API Schema
+    '''
+
+    assets = fields.List(fields.UUID())
+    tags = fields.List(fields.UUID())
+    action = fields.String(validate=v.OneOf(['add', 'remove']))
+
+
+class ImportAssetSchema(Schema):
+    '''
+    Import Asset API Schema
+    '''
+
+    assets = fields.List(fields.Dict(), validate=v.Length(min=1))
+    source = fields.String()
+
+    @pre_load
+    def transform_assets(
+        self, data, **kwargs
+    ):
+        '''
+        Transform a Tuple of Assets into a List of Assets
+        and validate each Asset.
+        '''
+        data['assets'] = list(data.get('assets', []))
+
+        if len(data['assets']) < 1:
+            raise ValidationError('Provide at least one asset')
+
+        for asset in data['assets']:
+            if not (
+                asset.get('fqdn')
+                or asset.get('ipv4')
+                or asset.get('netbios_name')
+                or asset.get('mac_address')
+            ):
+                raise ValidationError(
+                    'Each asset object requires a value for at least one '
+                    'of the following properties: '
+                    'fqdn, ipv4, netbios_name, mac_address.'
+                )
+        return data
```

Comment:
Shouldn't this be handled with an appropriate asset schema and validation method?

```python
class AssetSchema(Schema):
    class Meta:
        unknown = INCLUDE
    
    @validates_schema
    def validate_required_attributes_exist(self, data, **kwargs):
        if not (data.get('fqdn') 
                or data.get('ipv4') 
                or data.get('netbios_name') 
                or data.get('mac_address')
                ):
            raise ValidationError('Each asset object requires a value for at '
                                  'least one of the following properties: '
                                  'fqdn, ipv4, netbios_name, mac_address.'
                                  )


class ImportAssetSchema(Schema):
    assets = fields.List(fields.nested(AssetSchema), 
                         validate=v.Length(min=1),
                         required=True
                         )
    source = fields.Str(required=True)
```

Reply:
yes